### PR TITLE
Implement Feature: /colorcode (#56)

### PR DESCRIPTION
**Pull request by orbyfied**
Complete issue #56

### CODE REVIEW

**Comment by orbyfied** [comment]
dont merge yet

**Comment by orbyfied** [comment]
there are bugs

**Comment by orbyfied** [comment]
Its complete now, all bugs eliminated
@Matthias1590 you can merge

**Review by Matthias1590 — CHANGES_REQUESTED** [review]

File: src/main/java/com/domain/redstonetools/features/commands/colorcode/ColorCodeFeature.java [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+import net.minecraft.server.command.ServerCommandSource;
+import net.minecraft.server.network.ServerPlayerEntity;
+import net.minecraft.text.Style;
+import net.minecraft.text.Text;
+import net.minecraft.util.Formatting;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.concurrent.atomic.AtomicInteger;
+
+@Feature(name = "colorcode")
```
[the remark this answers is not in the document]
Comment (orbyfied) [reply src/main/java/com/domain/redstonetools/features/commands/colorcode/ColorCodeFeature.java:31]: In the document it says just `colorcode`, only r-stack has another `/` but sure.

File: src/main/java/com/domain/redstonetools/utils/CharTreeNode.java [hunk 2]
```diff
@@ -0,0 +1,95 @@
+package com.domain.redstonetools.utils;
```
[the remark this answers is not in the document]
Comment (orbyfied) [reply src/main/java/com/domain/redstonetools/utils/CharTreeNode.java:1]: This is much more performant than multiple `endsWith` calls, which is very useful as we are doing `endsWith` on a block ID with 5 strings every block in the selection every single time the command was run.

In my opinion, this optimization is relatively needed.
I'd prefer to keep it.

[the remark this answers is not in the document]
Comment (Matthias1590) [reply src/main/java/com/domain/redstonetools/utils/CharTreeNode.java:1]: A compiled regex should be even more performant, if performance is an issue I'd prefer to use regex here, making someone read through the entire tree class just to have them realize it's the same as String.endsWith doesn't sound like a great idea to me

[the remark this answers is not in the document]
Comment (orbyfied) [reply src/main/java/com/domain/redstonetools/utils/CharTreeNode.java:1]: I don't think a compiled regex will be faster here.
That's practically the same as doing 5 `endsWith` calls, as you need a separate group for each literal.
I have added comments to the class though.

**Comment by orbyfied** [comment]
@Matthias1590 I've implemented all changes except removing the CharTreeNode, which I documented, so it is not necessary to read through the source code anymore.

**Comment by Matthias1590** [comment]
Adding documentation doesn't mean you don't have to read the code, if you can show me that this class is a lot faster than regex we can use it tho I still don't like the unnecessary complexity it adds, readability is more important than performance in this case imo

**Comment by orbyfied** [comment]
I benchmarked it, ~70% faster than regex, but it doesn't really make a big difference in this case I think, so I moved to regex as you wished.

**Comment by orbyfied** [comment]
Actually now that I think about it, we might be doing this on millions of blocks at a time. If it gets noticeable I suggest we switch back to the char tree.

**Review by Matthias1590 — CHANGES_REQUESTED** [review]
The tree class and benchmark classes can be removed, other than that this looks great

**Comment by orbyfied** [comment]
Regex doesn't seem to be working.
The filter is completely broken.

**Comment by orbyfied** [comment]
I fixed the regex and removed all traces of CharTreeNode
@Matthias1590 You can merge now

**Review by Matthias1590 — APPROVED** [review]

**Pull request merged**